src/utils/editor/source-search.js

```diff
@@ -120,18 +129,31 @@ function updateCursor(cm, state, keepSelection) {
  */
 function doSearch(ctx, rev, query, keepSelection, modifiers: SearchModifiers) {
   let { cm } = ctx;
+  let matchIndex;
   cm.operation(function() {
     if (!query || isWhitespace(query)) {
       return;
     }
 
-    let state = getSearchState(cm);
+    let state = getSearchState(cm, query, modifiers);
     state.query = query;
 
     updateOverlay(cm, state, query, modifiers);
     updateCursor(cm, state, keepSelection);
-    searchNext(ctx, rev, modifiers);
+
+    const nextMatch = searchNext(ctx, rev, query, modifiers);
+
+    if (nextMatch) {
+      if (state.matchIndex === -1) {
+        state.matchIndex = findIndex(state.results, nextMatch);
+      } else {
+        state.matchIndex = rev ? state.matchIndex - 1 : state.matchIndex + 1;
+      }
+      matchIndex =
+        (state.matchIndex + state.results.length) % state.results.length;
+    }
```

Comment:
Hmm, looks like this is giving us a weird value... and it can likely be cleaned up further.

I'm going to try moving the modulo math to the increment / decrement line

<img width="409" alt="screen shot 2017-02-17 at 3 27 31 pm" src="https://cloud.githubusercontent.com/assets/254562/23082135/17f34fce-f526-11e6-8555-a64d613a13f1.png">